### PR TITLE
feat: [CDS-74583]:Fixed-issue-with-empty-variables-field

### PR DESCRIPTION
**Pull request by litvinovstas**
### Summary
Fixed issue with unsaved empty array for variables in stage. in overview tab advanced section.

#### Screenshots

NA

#### PR Checklist

<details>
<summary>Please check if your PR fulfils the following requirements</summary>

- [ ] Tests for the changes have been added. Ideally, include a test that fails without this PR but passes with it.
- [ ] Docs have been [added/updated](https://harness.atlassian.net/jira/software/c/projects/DOC/boards/40).
</details>

<details>
<summary>Use the following comments to re-trigger PR Checks</summary>

- Jest: `retrigger jest`
- Prettier: `retrigger prettier`
- Type Check: `retrigger typecheck`
- ESLint: `retrigger eslint`
- Sonar: `retrigger sonar`
- Standards: `retrigger standards`
- Build: `retrigger build`
- Title Check: `retrigger titlecheck`
- Feature Name Check: `trigger featurenamecheck`
- Coverage: `retrigger coverage`
- Rebase: `trigger rebase`
- Cypress Rest: `retrigger cypress-rest`
- Cypress CD: `retrigger cypress-cd`
- Cypress Pipeline: `retrigger cypress-pipeline`
- Cypress CV: `retrigger cypress-cv`
- Fix Prettier: `fix prettier`
</details>

### CODE REVIEW

File: src/modules/70-pipeline/components/PipelineSteps/Steps/CustomVariables/CustomVariables.tsx [hunk 1]
```diff
@@ -190,24 +192,27 @@ export class CustomVariables extends Step<CustomVariablesData> {
...
+            ...(!isNil(row.description) ? { description: row.description } : {}),
+            ...(!isNil(row.required) ? { required: row.required } : {}),
+            value:
+              row.type === 'Number' &&
+              getMultiTypeFromValue(row.value as unknown as string) === MultiTypeInputType.FIXED &&
+              row.value
+                ? parseFloat(row.value as unknown as string)
+                : row.value
+          })) as AllNGVariables[])
+        : (undefined as unknown as AllNGVariables[])
```
Comment (rameshx) [inline src/modules/70-pipeline/components/PipelineSteps/Steps/CustomVariables/CustomVariables.tsx:215]: This doesn't look right. If `variables` can be undefined, we should update `CustomVariablesData` to make `variables` optional.

**Review by rameshx — CHANGES_REQUESTED** [review]

File: src/modules/70-pipeline/components/PipelineSteps/Steps/CustomVariables/CustomVariables.tsx [hunk 2]
```diff
@@ -28,6 +28,7 @@ import type { StepProps } from '../../PipelineStep'
 
 const logger = loggerFor(ModuleName.COMMON)
 
+/* istanbul ignore next */
```
[the remark this answers is not in the document]
Comment (litvinovstas) [reply src/modules/70-pipeline/components/PipelineSteps/Steps/CustomVariables/CustomVariables.tsx:31]: it should be covered by jest in separated pr.

File: src/modules/70-pipeline/components/PipelineSteps/Steps/CustomVariables/CustomVariables.tsx (hunk 1 above)
Comment (litvinovstas) [reply src/modules/70-pipeline/components/PipelineSteps/Steps/CustomVariables/CustomVariables.tsx:215]: improved

**Review by satyam-harness — DISMISSED** [review]

**Review by rameshx — DISMISSED** [review]

**Review by Sagar-Rajput7 — APPROVED** [review]

File: src/modules/70-pipeline/components/PipelineSteps/Steps/CustomVariables/CustomVariableEditable.tsx [hunk 3]
```diff
@@ -214,13 +214,14 @@ export function CustomVariableEditable(props: CustomVariableEditableProps): Reac
                 {headerComponent
                   ? headerComponent
                   : props.showHeaders &&
-                    values.variables.length > 0 && (
+                    values?.variables &&
+                    values.variables?.length > 0 && (
```
Comment (d1wilko) [inline src/modules/70-pipeline/components/PipelineSteps/Steps/CustomVariables/CustomVariableEditable.tsx:218]: not an issue but I think you could get away with `values?.variables?.length && (` here

File: src/modules/70-pipeline/components/PipelineSteps/Steps/CustomVariables/CustomVariableInputSet.tsx [hunk 4]
```diff
@@ -116,7 +116,7 @@ function CustomVariableInputSetBasic(props: ConectedCustomVariableInputSetProps)
 
   return (
     <div className={cx(css.customVariablesInputSets, 'customVariables', className)} id={domId}>
-      {initialValues.variables.length > 0 && (
+      {initialValues?.variables && initialValues.variables?.length > 0 && (
```
Comment (d1wilko) [inline src/modules/70-pipeline/components/PipelineSteps/Steps/CustomVariables/CustomVariableInputSet.tsx:119]: same as above but not worth holding up the PR for

File: src/modules/70-pipeline/components/PipelineSteps/Steps/CustomVariables/CustomVariablesEditableStage.tsx [hunk 5]
```diff
@@ -155,14 +155,14 @@ export function CustomVariablesEditableStage(props: CustomVariableEditableProps)
                     formName={formName}
                     allowedVarialblesTypes={allowedVarialblesTypes}
                   />
-                  {values.variables?.length > 0 ? (
+                  {values?.variables && values.variables?.length > 0 ? (
```
Comment (d1wilko) [inline src/modules/70-pipeline/components/PipelineSteps/Steps/CustomVariables/CustomVariablesEditableStage.tsx:158]: same as above

File: src/modules/75-ci/components/PipelineStudio/BuildStageSpecifications/BuildStageSpecifications.tsx [hunk 6]
```diff
@@ -221,7 +221,7 @@ export default function BuildStageSpecifications({ children }: React.PropsWithCh
 
         spec = set(spec, 'caching.key', values.cacheIntelligenceKey)
 
-        if (values.variables && values.variables.length > 0) {
+        if (values?.variables && values.variables?.length > 0) {
```
Comment (d1wilko) [inline src/modules/75-ci/components/PipelineStudio/BuildStageSpecifications/BuildStageSpecifications.tsx:224]: same as above

**Review by d1wilko — APPROVED** [review]
some suggestions but not worth holding up the PR for :)

**Review by rameshx — APPROVED** [review]

**Review by satyam-harness — APPROVED** [review]

**Pull request merged**